### DISCUSSION TITLE
Deprecated: File Map Storage: How To and Q&A

### DISCUSSION THREAD

**Discussion by hmlnarik** [keycloak/keycloak#16768] [Archive]
## Setup Keycloak server with files as a storage

Currently available in #16001

To run, execute the following command:

```
mvn -pl testsuite/utils exec:java \
 -Pkeycloak-server,map-storage \
 -Dkeycloak.mapStorage.provider.default=file \
 -Dkeycloak.singleUseObject.map.storage.provider=concurrenthashmap
```

You can now also use nightly Quarkus distribution:

```
bin/kc.sh start-dev --storage=file --storage-file-dir=data/file
```

### Known issues and limitations
Summary:
- No-downtime upgrade not supported yet
- File locking not supported yet
- Single-use store not applicable to file store, currently uses `chm` store is configured instead (other implementation can be used via `--storage-area-single-use-object` option)
- Format of the files is in alpha stage and will change
- No caching in place, thus file store is slow

[Open issues](https://github.com/orgs/keycloak/projects/9/views/1?filterQuery=is%3Aissue++-status%3A%22%E2%9C%85+Done%22+new-store-scope%3A%22%F0%9F%94%A0+File+store+%28physical+layer%29%22)

**Comment by dominikschlosser** [comment]
We tested it and it works without any problems.

We want to use this store for mostly static data (realms, clients, clientscopes) in a configuration-as-code like manner. Right now we are using a third-party tool (Config-CLI) to provision a running Keycloak instance with configuration data from YAML files.

Some things that are bothering us are:
- The file and directory structure of current file storage can't be used easily in Kubernetes since ConfigMaps can only represent one directory. For realms, clients, clientscopes and roles we currently need to define 14 configmaps for 3 realms.
It would be nice if there was an option to put everything belonging to a realm in single file.
- Currently there seems to be no way to substitute environment variables inside the configuration. This is necessary for secrets like private keys, which are provided as Kubernetes-secrets and exposed as environment variables inside the container.

**Reply by ahus1** [reply]
Thanks for sharing your experiences. About the file structure: While I can't promise any changes today, as an intermediate solution considering zip'ing the file storage, putting it into a ConfigMap, and unzip'ing it either at Keycloak startup, or using a sidecar.

**Comment by kereis** [comment]
I was redirected to here by mailing list as this feature might be relevant for my use case. 

When I look at #16001, the handling of authz related settings hasn't been implemented yet. Is there a possible ETA when this will be implemented? If there's no plan to implementing it in the near future, is it possible to contribute to this feature?

**Reply by hmlnarik** [reply]
Thank you very much for your interest and offer!

Support for authz is improved in [a private branch](https://github.com/hmlnarik/keycloak/tree/file-store-authz) which should be gradually migrated to main. If you are interested and able to try it, I would warmly appreciate your feedback.

**Reply by hmlnarik** [reply]
The authz has been merged to the main branch today and should be ready to be tested in the next nightly. If you are still interested in trying it out, we would welcome your feedback!

**Comment by kereis** [comment]
I haven't looked enough into detail, but how does the new file map storage handle changes in static files? Does a change recreate the affected entity or update only the changed properties?

My use case here: We have configured a user federation in `READ_ONLY` mode, meaning that any user-specific information, e.g. role mappings, are saved in Keycloak DB only. If a realm file is updated and thus leading to the realm being recreated, user role mappings will disappear.

**Reply by hmlnarik** [reply]
Thank you for the question. The file storage stores the data only in the file system, so any change in static files is immediately reflected by Keycloak.

Please beware that the map storage does not work with the user federation at this moment.

**Comment by sjhr007** [comment]
How can stored data files be managed in a versioning system (like Git)? Does Keycloak provide ways to commit files to a versioning system when files get updated through operations from Keycloak UI? 
How can changes be propagated to a versioning system like updating realm or client definition/attributes from Keycloak UI updates files in the filesystem?

**Comment by ahus1** [comment]
As announced to the community in [this blog post](https://www.keycloak.org/2023/10/map-store-removal), we are no longer developing the Map Store, and it will be removed from Keycloak's codebase. Follow #24111 for the removal steps.
